### PR TITLE
plan: add cast function for PatternInExpr.

### PR DESCRIPTION
**Pull request by coocood**
Fixes a bug when pattern in items input is string but pattern column is of integer type.

### CODE REVIEW

File: util/types/datum.go [hunk 1]
```diff
@@ -1251,10 +1203,11 @@ func (d *Datum) ToInt64() (int64, error) {
 	case KindString:
 		s := d.GetString()
 		fval, err := StrToFloat(s)
-		if err != nil {
-			return 0, errors.Trace(err)
+		i64, err1 := convertFloatToInt(fval, lowerBound, upperBound, tp)
+		if err == nil {
```
Comment (hanfei1991) [inline util/types/datum.go]: else {
 err = errors.Trace(err)
}

**Comment by hanfei1991** [comment]
LGTM

File: plan/typeinferer.go [hunk 2]
```diff
@@ -373,3 +376,33 @@ func (v *typeInferrer) handleCaseExpr(x *ast.CaseExpr) {
...
+				}
+				if cmp != 0 {
+					// The value will never match the column, do not set newDatum.
+					continue
+				}
+				valueExpr.SetDatum(newDatum)
+			}
+		}
+		if v.err != nil {
+			// TODO: errors should be handled differently according to query context.
```
Comment (zimulala) [inline plan/typeinferer.go]: s/errors/Errors

File: plan/typeinferer.go [hunk 3]
```diff
@@ -373,3 +376,33 @@ func (v *typeInferrer) handleCaseExpr(x *ast.CaseExpr) {
 	// TODO: We need a better way to set charset/collation
 	x.Type.Charset, x.Type.Collate = types.DefaultCharsetForType(x.Type.Tp)
 }
+
+// ConvertValueToColumnTypeIfNeeded checks if the expr in PatternInExpr is column name,
+// and cast function to the items in the list.
```
Comment (zimulala) [inline plan/typeinferer.go]: s/cast/casts

**Comment by zimulala** [comment]
LGTM

**Pull request merged**